### PR TITLE
Optimize decimal state serializers for small value case

### PR DESCRIPTION
**Pull request by lukasz-stec**
## Description

Given that many decimal aggregations (sum, avg) stay in the `long` range, aggregation state serializer can be optimized for this case, limiting the number of bytes per position significantly (3-4X).

#### tpch/tpcds benchmarks
image

[Benchmarks_decimal_aggr_serde_simple_case.pdf](https://github.com/trinodb/trino/files/9291809/Benchmarks_decimal_aggr_serde_simple_case.pdf)

> Is this change a fix, improvement, new feature, refactoring, or other?

improvement
> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

core query engine (`sum`, `avg` aggregation state serialization)
> How would you describe this change to a non-technical end user or system administrator?

improve performance of queries with `sum` or `avg` aggregations
## Documentation

( X) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(X ) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java [hunk 1]
```diff
@@ -42,7 +42,13 @@ public void serialize(LongDecimalWithOverflowAndLongState state, BlockBuilder ou
             long overflow = state.getOverflow();
             long[] decimal = state.getDecimalArray();
             int offset = state.getDecimalArrayOffset();
-            VARBINARY.writeSlice(out, Slices.wrappedLongArray(count, overflow, decimal[offset], decimal[offset + 1]));
+            if (count == 1 && overflow == 0 && decimal[offset] == 0) {
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java:45]: `count == 1 && overflow == 0` should be separate from `decimal[offset] == 0`

`decimal[offset] == 0` will be true often when `count == 1 && overflow == 0` isn't.

So you have cases:
* 1 long -> count == 1 && overflow == 0 && decimal[offset] == 0
* 2 longs -> count == 1 && overflow == 0
* 3 longs -> decimal[offset] == 0
* 4 longs -> full case

I think you can even make it branchless:
* append `decimal[offset + 1]` unconditionally, len += 1
* append `decimal[offset]` unconditionally, len = len + (decimal[offset] == 0 ? 0 : 1)
* append  `overflow` and `count`, len = len + ((count == 1 & overflow == 0) ? 0 : 2)

File: core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java [hunk 2]
```diff
@@ -54,20 +60,31 @@ public void deserialize(Block block, int index, LongDecimalWithOverflowAndLongSt
...
+            if (slice.length() == Long.BYTES) {
+                state.setLong(1);
+                state.setOverflow(0);
+                state.setNotNull();
+                long[] decimal = state.getDecimalArray();
+                int offset = state.getDecimalArrayOffset();
+                decimal[offset] = 0;
+                decimal[offset + 1] = slice.getLong(0);
             }
+            else {
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java:72]: similar comment as in https://github.com/trinodb/trino/pull/13573/files#r942932796 but reverse

File: core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java (hunk 1 above)
Comment (lukasz-stec) [reply core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java:45]: this may help in some cases, I will try that. This probably won't help in tpch q17/ q18 as it's already hitting the condition.

**Comment by lukasz-stec** [comment]
Made it draft again as I'm gonna benchmark supporting more cases

File: core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java [hunk 3]
```diff
@@ -42,7 +42,14 @@ public void serialize(LongDecimalWithOverflowAndLongState state, BlockBuilder ou
             long overflow = state.getOverflow();
             long[] decimal = state.getDecimalArray();
             int offset = state.getDecimalArrayOffset();
-            VARBINARY.writeSlice(out, Slices.wrappedLongArray(count, overflow, decimal[offset], decimal[offset + 1]));
+            long[] buffer = new long[4];
+            buffer[0] = decimal[offset + 1];
+            buffer[1] = decimal[offset];
+            buffer[2] = overflow;
+            buffer[3] = count;
+            int bufferLength = 1 + (decimal[offset] == 0 ? 0 : 1);
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java:50]: `decimal[offset] == 0 ? 0 : 1` -> `decimal[offset] ^ 0` (please check if that is correct)

Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java:50]: bufferLength -> decimalsCount ?

File: core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java [hunk 4]
```diff
@@ -53,21 +60,25 @@ public void serialize(LongDecimalWithOverflowAndLongState state, BlockBuilder ou
...
-            if (slice.length() != SERIALIZED_SIZE) {
-                throw new IllegalStateException("Unexpected serialized state size: " + slice.length());
-            }
-
-            long count = slice.getLong(0);
-            long overflow = slice.getLong(Long.BYTES);
-
-            state.setLong(count);
-            state.setOverflow(overflow);
             state.setNotNull();
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java:63]: Maybe move this to the end to keep all the state updates together

File: core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowStateSerializer.java [hunk 5]
```diff
@@ -41,7 +41,13 @@ public void serialize(LongDecimalWithOverflowState state, BlockBuilder out)
             long overflow = state.getOverflow();
             long[] decimal = state.getDecimalArray();
             int offset = state.getDecimalArrayOffset();
-            VARBINARY.writeSlice(out, Slices.wrappedLongArray(overflow, decimal[offset], decimal[offset + 1]));
+            long[] buffer = new long[3];
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowStateSerializer.java:44]: Similar comments

**Review by raunaqmorarka — COMMENTED** [review]
Is `BenchmarkBlockSerde#serializeLongDecimal` or any other other JMH benchmark relevant here ?

**Review by lukasz-stec — COMMENTED** [review]
Comments addressed

**Comment by lukasz-stec** [comment]
>Is BenchmarkBlockSerde#serializeLongDecimal or any other other JMH benchmark relevant here ?

@raunaqmorarka I think `BenchmarkDecimalAggregation.benchmarkEvaluateIntermediate` is the best match for those changes although the first step and 50% of the benchmark time is the `addInput`.

I don't see many changes in this microbenchmark. The important thing to notice is that the number of groups thus random memory load times is the most important factor here.
```
Before
Benchmark                                                  (function)  (groupCount)  (type)  Mode  Cnt   Score   Error  Units
BenchmarkDecimalAggregation.benchmarkEvaluateIntermediate         avg          1000    LONG  avgt   20   7.190 ± 0.307  ns/op
BenchmarkDecimalAggregation.benchmarkEvaluateIntermediate         avg       1000000    LONG  avgt   20  99.130 ± 1.451  ns/op

After
Benchmark                                                  (function)  (groupCount)  (type)  Mode  Cnt   Score   Error  Units
BenchmarkDecimalAggregation.benchmarkEvaluateIntermediate         avg          1000    LONG  avgt   20   6.856 ± 0.037  ns/op
BenchmarkDecimalAggregation.benchmarkEvaluateIntermediate         avg       1000000    LONG  avgt   20  94.083 ± 1.077  ns/op
```

File: core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java (hunk 3 above)
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java]: I meant that only for the `1 + (decimal[offset] == 0 ? 0 : 1)` part, the variable on next line should definitely be named `bufferLength`.

File: core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowStateSerializer.java [hunk 6]
```diff
@@ -53,18 +64,22 @@ public void deserialize(Block block, int index, LongDecimalWithOverflowState sta
...
+            long overflow = 0;
+            if (slice.length() > Long.BYTES) {
+                high = slice.getLong(Long.BYTES);
+                if (slice.length() == SERIALIZED_SIZE) {
+                    overflow = slice.getLong(Long.BYTES * 2);
+                }
+            }
 
-            state.setOverflow(overflow);
             state.setNotNull();
```
Comment (raunaqmorarka) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowStateSerializer.java:79]: nit: move this to last line to keep similar flow as before

File: core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java (hunk 3 above)
Comment (raunaqmorarka) [reply core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java]: ya, that doesn't work

**Review by raunaqmorarka — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java (hunk 3 above)
Comment (lukasz-stec) [reply core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java]: that's actually way better. Renamed.

File: core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java [hunk 7]
```diff
@@ -42,7 +42,19 @@ public void serialize(LongDecimalWithOverflowAndLongState state, BlockBuilder ou
...
             int offset = state.getDecimalArrayOffset();
-            VARBINARY.writeSlice(out, Slices.wrappedLongArray(count, overflow, decimal[offset], decimal[offset + 1]));
+            long[] buffer = new long[4];
+            long decimalLowBytes = decimal[offset + 1];
+            long decimalHighBytes = decimal[offset];
+            buffer[0] = decimalLowBytes;
+            buffer[1] = decimalHighBytes;
+            buffer[2] = overflow;
+            buffer[3] = count;
+            // if decimalHighBytes == 0 and count == 1 and overflow == 0 we only write decimalLowBytes (bufferLength = 1)
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java:52]: You miss one important case: ```3 longs -> decimal[offset] == 0```
see https://github.com/trinodb/trino/pull/13573#discussion_r942932796

Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java:52]: You can implement it by:
```
// append low
buffer[0] = decimalLowBytes;
offset = 1;
// append high
buffer[offset] = decimalHiBytes;
offset = offset + (decimalHiBytes == 0? 0 : 1);
// append overflow, count
buffer[offset] = buffer[offset] | overflow;
buffer[offset + 1] = count;
offset = offset + (overflow == 0 & count == 1) ? 0 : 2
```

File: core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java [hunk 8]
```diff
@@ -54,20 +66,25 @@ public void deserialize(Block block, int index, LongDecimalWithOverflowAndLongSt
...
-            }
+            long[] decimal = state.getDecimalArray();
+            int offset = state.getDecimalArrayOffset();
 
-            long count = slice.getLong(0);
-            long overflow = slice.getLong(Long.BYTES);
+            long high = 0;
+            long overflow = 0;
+            long count = 1;
+            if (slice.length() > Long.BYTES) {
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java:75]: branchless version
```
int sliceLength = slices.length();
long low = slice.getLong(0);
long highOffset = sliceLength == 4 | sliceLength == 2 ? 1 : 0; 
long high = slice.getLong(highOffset) & (highOffset * -1L);
// similar for count & overflow
```
Would be great to learn if JIT actually generates branchless versions

File: core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java [hunk 9]
```diff
@@ -42,7 +42,19 @@ public void serialize(LongDecimalWithOverflowAndLongState state, BlockBuilder ou
             long overflow = state.getOverflow();
             long[] decimal = state.getDecimalArray();
             int offset = state.getDecimalArrayOffset();
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java:44]: Could we actually have UT for serde?

**Comment by lukasz-stec** [comment]
the latest version tpch/tpcds benchmarks

good improvement for tpch
image

[Benchmarks_decimal_aggr_serde_opt.pdf](https://github.com/trinodb/trino/files/9374630/Benchmarks_decimal_aggr_serde_opt.pdf)

**Review by lukasz-stec — COMMENTED** [review]
I implemented the additional case + UT.
For the branchless deserialization. I tested multiple ways (including offered by @sopel39 ) and it there is a little difference in jmh for mixed decimals cases (block with both small and large decimals) but if/switch approach is a little better for both small and large cases where the branch prediction works efficiently.

The tpc benchmarks look very good with the current code (switch).

**Comment by lukasz-stec** [comment]
`tpch/tpcds orc part 1k` on latest version (missing case implemented + using switch in deserailziation).
Big gains for tpch, some for tpcds.

image

[Benchmarks_decimal_aggr_serde_opt_orc_part_sf1k_250822.pdf](https://github.com/trinodb/trino/files/9422673/Benchmarks_decimal_aggr_serde_opt_orc_part_sf1k_250822.pdf)

File: core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java [hunk 10]
```diff
@@ -53,12 +54,12 @@
 @State(Scope.Thread)
 @OutputTimeUnit(TimeUnit.NANOSECONDS)
 @Fork(3)
-@Warmup(iterations = 10)
-@Measurement(iterations = 10)
+@Warmup(iterations = 20, time = 1)
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java:57]: nit: is time needed?

File: core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java [hunk 11]
```diff
@@ -82,6 +83,27 @@ public Block benchmarkEvaluateIntermediate(BenchmarkData data)
...
+        BlockBuilder builder = aggregator.getType().createBlockBuilder(null, data.getGroupCount());
+        for (int groupId = 0; groupId < data.getGroupCount(); groupId++) {
+            aggregator.evaluate(groupId, builder);
+        }
+        return builder.build();
+    }
+
+    @Benchmark
+    @OutputTimeUnit(TimeUnit.MICROSECONDS)
+    public Object benchmarkDeserialize(BenchmarkData data)
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java:100]: Why it's called `benchmarkDeserialize`? I think `processPage` does more than just deserialization

File: core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java [hunk 12]
```diff
@@ -108,36 +130,66 @@ public Block benchmarkEvaluateFinal(BenchmarkData data)
         @Param({"10", "1000"})
         private int groupCount = 10;
 
+        @Param({"SHORT", "LONG", "MIXED"})
+        private String decimalSize = "SHORT";
+
+        @Param({"true", "false"})
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java:136]: How does group ordering matter?

File: core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java [hunk 13]
```diff
@@ -39,6 +39,7 @@
 import org.testng.annotations.Test;
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java:39]: this should be separate PR

File: core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java [hunk 14]
```diff
@@ -108,36 +130,66 @@ public Block benchmarkEvaluateFinal(BenchmarkData data)
         @Param({"10", "1000"})
         private int groupCount = 10;
 
+        @Param({"SHORT", "LONG", "MIXED"})
+        private String decimalSize = "SHORT";
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java:134]: decimalSize=`LONG` won't work with `type=SHORT`

File: core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java [hunk 15]
```diff
@@ -108,36 +130,66 @@ public Block benchmarkEvaluateFinal(BenchmarkData data)
...
-
+            Random random = new Random(343526534);
             switch (type) {
                 case "SHORT": {
                     DecimalType type = createDecimalType(14, 3);
-                    values = createValues(functionResolution, type, type::writeLong);
+                    values = createValues(functionResolution, type, (builder, value) -> {
+                        boolean writeShort = "SHORT".equals(decimalSize) || ("MIXED".equals(decimalSize) && random.nextBoolean());
+                        if (writeShort) {
+                            builder.writeLong(value);
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java:157]: both cases are long really `writeLong`

File: core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java [hunk 16]
```diff
@@ -42,7 +39,29 @@ public void serialize(LongDecimalWithOverflowAndLongState state, BlockBuilder ou
...
-            VARBINARY.writeSlice(out, Slices.wrappedLongArray(count, overflow, decimal[offset], decimal[offset + 1]));
+            long[] buffer = new long[4];
+            long high = decimal[offset];
+            long low = decimal[offset + 1];
+
+            // append low
+            buffer[0] = low;
+            // append low
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowAndLongStateSerializer.java:48]: `append high`

File: core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowStateSerializer.java [hunk 17]
```diff
@@ -41,7 +38,18 @@ public void serialize(LongDecimalWithOverflowState state, BlockBuilder out)
...
+            long low = decimal[offset + 1];
+            long high = decimal[offset];
+            buffer[0] = low;
+            buffer[1] = high;
+            buffer[2] = overflow;
+            // if high == 0 and overflow == 0 we only write low (bufferLength = 1)
+            // if high != 0 and overflow == 0 we write both low and high (bufferLength = 2)
+            // if overflow != 0 we write all values (bufferLength = 3)
+            int decimalsCount = 1 + (high == 0 ? 0 : 1);
+            int bufferLength = overflow == 0 ? decimalsCount : 3;
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowStateSerializer.java:51]: I think `overflow == 0 ? decimalsCount : 3;` might actually turn into branch

File: core/trino-main/src/test/java/io/trino/operator/aggregation/state/TestLongDecimalWithOverflowAndLongStateSerializer.java [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+        serializer.serialize(state, out);
+
+        Block serialized = out.build();
+        LongDecimalWithOverflowAndLongState outState = STATE_FACTORY.createSingleState();
+        serializer.deserialize(serialized, 0, outState);
+        return outState;
+    }
+
+    @DataProvider
+    public Object[][] input()
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/operator/aggregation/state/TestLongDecimalWithOverflowAndLongStateSerializer.java:74]: please add `expectedLength` to each row

File: core/trino-main/src/test/java/io/trino/operator/aggregation/state/TestLongDecimalWithOverflowAndLongStateSerializer.java [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+    {
+        return new Object[][] {
+                {1, 0, 0, 1},
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/operator/aggregation/state/TestLongDecimalWithOverflowAndLongStateSerializer.java:77]: you should use different values than `1` for each column to make sure two values are not mismatched

**Review by sopel39 — COMMENTED** [review]
lgtm `[Optimize decimal state serializers for small value case](https://github.com/trinodb/trino/pull/13573/commits/18708885459f1921b720148552a14d1ca8ff7c93)` % comments

**Comment by sopel39** [comment]
Nice gains

File: core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowStateSerializer.java (hunk 17 above)
Comment (lukasz-stec) [reply core/trino-main/src/main/java/io/trino/operator/aggregation/state/LongDecimalWithOverflowStateSerializer.java:51]: it may be a branch, but usually well predicted.Even in the mixed case I don't see large impact
```

Benchmark                                                      (decimalSize)  (function)  (groupCount)  (groupIdsInOrder)  (type)  Mode  Cnt      Score     Error  Units
BenchmarkDecimalAggregation.benchmarkEvaluateIntermediateOnly          SHORT         sum       1000000               true    LONG  avgt   10  36281.881 ± 834.705  us/op
BenchmarkDecimalAggregation.benchmarkEvaluateIntermediateOnly           LONG         sum       1000000               true    LONG  avgt   10  56961.139 ± 372.546  us/op
BenchmarkDecimalAggregation.benchmarkEvaluateIntermediateOnly          MIXED         sum       1000000               true    LONG  avgt   10  52997.475 ± 314.941  us/op
```

File: core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java (hunk 13 above)
Comment (lukasz-stec) [reply core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java:39]: all changes to `BenchmarkDecimalAggregation`?

File: core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java (hunk 10 above)
Comment (lukasz-stec) [reply core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java]: yeah, 1s is not enough to trigger c2.  5s warmup does the trick

**Review by lukasz-stec — COMMENTED** [review]
comments addressed

File: core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java (hunk 11 above)
Comment (sopel39) [reply core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java:100]: Still, for a code reader this is confusing.

> there is no way to benchmark deserialize only so

Sure there is: `LongDecimalWithOverflowAndLongStateSerializer` is a class and there is nothing stopping us from benchmarking it directly

File: core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java (hunk 14 above)
Comment (sopel39) [reply core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java:134]: > in this case, I use larger long values (close to Long.MAX_VALUE)

It doesn't matter from perf perspective.

File: core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java (hunk 12 above)
Comment (sopel39) [reply core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java:136]: > For the partial aggregation adaptation case, the group ids are always consecutive.

This has little to do with serialization.

> It doesn't matter for serialization, but for evaluateIntermediate the main bottleneck is actually jumping to random memory locations in case groups are located randomly.

For skip-cases groups are consecutive and number of them is small, so it probably doesn't matter too.

File: core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java (hunk 15 above)
Comment (sopel39) [reply core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java:157]: > yes, but if we use varint those are different cases.

Here `SHORT` vs `LONG` is mixed with Short/Long decimal. This is confusing

File: core/trino-main/src/test/java/io/trino/operator/aggregation/state/TestLongDecimalWithOverflowAndLongStateSerializer.java (hunk 18 above)
Comment (sopel39) [reply core/trino-main/src/test/java/io/trino/operator/aggregation/state/TestLongDecimalWithOverflowAndLongStateSerializer.java:74]: no, it checks that `LongDecimalWithOverflowAndLongStateSerializer` encoded data correctly and efficiently.

**Review by sopel39 — COMMENTED** [review]
reviewed again

File: core/trino-main/src/test/java/io/trino/operator/aggregation/state/TestLongDecimalWithOverflowAndLongStateSerializer.java (hunk 18 above)
Comment (lukasz-stec) [reply core/trino-main/src/test/java/io/trino/operator/aggregation/state/TestLongDecimalWithOverflowAndLongStateSerializer.java:74]: `expectedLength` does not matter for the correctness

File: core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java (hunk 15 above)
Comment (lukasz-stec) [reply core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java:157]: replaced SHORT, LONG with SMALL, BIG

File: core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java (hunk 13 above)
Comment (lukasz-stec) [reply core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java:39]: @sopel39 ping

**Review by lukasz-stec — COMMENTED** [review]
comments addressed

File: core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java (hunk 13 above)
Comment (lukasz-stec) [reply core/trino-main/src/test/java/io/trino/operator/aggregation/BenchmarkDecimalAggregation.java]: extracted `BenchmarkDecimalAggregation` changes to https://github.com/trinodb/trino/pull/13939

**Review by sopel39 — APPROVED** [review]
